numpyro/distributions/continuous.py

```diff
@@ -1188,6 +1192,141 @@ def infer_shapes(
+        :class:`scipy.sparse.spmatrix`)
+    """
+
+    arg_constraints = {
+        "loc": constraints.real_vector,
+        "alpha": constraints.open_interval(-1, 1),
+        "tau": constraints.positive,
+        "W": constraints.dependent_property(is_discrete=False, event_dim=2),
+    }
+    support = constraints.real_vector
+    reparametrized_params = [
+        "loc",
+        "alpha",
+        "tau",
+        "W",
+    ]
+
+    def __init__(self, loc, alpha, tau, W, *, is_sparse=False, validate_args=None):
+        if jnp.ndim(loc) == 0:
+            (loc,) = promote_shapes(loc, shape=(1,))
+
+        self.is_sparse = is_sparse
+
+        if self.is_sparse:
+            if not sparse.issparse(W):
+                assert isinstance(W, np.ndarray)
+            self.W = sparse.csr_matrix(W)
+        else:
+            assert not sparse.issparse(W)
+            self.W = W
+
+        batch_shape = lax.broadcast_shapes(
+            jnp.shape(loc)[:-1],
+            jnp.shape(alpha),
+            jnp.shape(tau),
+            jnp.shape(self.W)[:-2],
+        )
+        event_shape = jnp.shape(self.W)[-1:]
+        (self.loc,) = promote_shapes(loc, shape=batch_shape + event_shape)
+        self.alpha, self.tau = promote_shapes(alpha, tau, shape=batch_shape)
+
+        super(CAR, self).__init__(
+            batch_shape=batch_shape,
+            event_shape=event_shape,
+            validate_args=validate_args,
+        )
+
+        if self._validate_args:
+            assert (
+                self.W.sum(axis=-1) > 0
+            ), "all sites in adjacency matrix W must have neighbours"
+
+            if sparse.issparse(self.W):
+                assert (self.W != self.W.T).nnz == 0, "W must be symmetric"
+            else:
+                assert np.array_equal(self.W, self.W.T), "W must be symmetric"
+
+    def sample(self, key, sample_shape=()):
+        # TODO: look into a sparse sampling method
+        mvn = MultivariateNormal(self.mean, precision_matrix=self.precision_matrix)
+        return mvn.sample(key, sample_shape=sample_shape)
+
+    @validate_sample
+    def log_prob(self, value):
+        phi = value - self.loc
+        W = self.W
+
+        if self.is_sparse:
+            D = np.asarray(W.sum(axis=-1)).squeeze()
+            D_rsqrt = D ** (-0.5)
+
+            W_scaled = W.multiply(D_rsqrt).multiply(D_rsqrt[:, np.newaxis]).toarray()
+
+            lam = np.linalg.eigvalsh(W_scaled)
+
+            W = BCOO.from_scipy_sparse(W)
+
+        else:
+            D = W.sum(axis=-1)
+            D_rsqrt = D ** (-0.5)
+
+            W_scaled = W * (D_rsqrt * D_rsqrt[:, jnp.newaxis])
+
+            lam = jnp.linalg.eigvalsh(W_scaled)
+
+        n = D.shape[-1]
+
+        logtau = n * jnp.log(self.tau)
+        logdet = jnp.log1p(-jnp.expand_dims(self.alpha, -1) * lam).sum(-1)
+
+        logquad = self.tau * jnp.sum(
+            phi * (D * phi - jnp.expand_dims(self.alpha, -1) * W @ phi), -1
```

Comment:
not sure if it helps, but you might want to do `(W @ phi)`

Reply:
I think the current version is mathematically correct. Would the brackets speed anything up?

Reply:
I thought it might resolve the issue you mentioned with multiplication for jax sparse matrix. I don't actually know the order of the operators here so the parentheses might be helpful to see that matmul is performed before taking the multiplication. Not sure about the speed though. If you want to perform multiplication first then you'll need to expand alpha dims by (-2,-1). Given a batch of scalar numbers a, the multiplication of a and a matrix M is `a[...,:.:] * M`.

Reply:
multiplication is fine for jax sparse matrix. I was confused because there is no `.multiply`, but `@` and `*` work fine.

I _think_ the line is fine as it is. Fewer paretheses make it more readable and scalar * W is still a sparse jax matrix

Reply:
I think you need to expand alpha dims to `jnp.expand_dims(alpha, (-2, -1))` if we want to multiply first, but I might be wrong. Let's see what happens in tests.

> Would the brackets speed anything up?

Just think that performing matrix-vector would be better in case we have a batch of `alpha`. Doing `alpha * W` first will create a batch of matrices, which makes the follow-up matrix-vector product much more expensive. ;)

Reply:
Seems like we need (W @ phi[..., None]).squeeze(axis=-1) here in case `phi` has batch dimension.

Reply:
Sorry I only just saw the comment above. This completely makes sense:

> In dense case, even for scalar alpha, doing (alpha * W) @ phi would be more expensive than alpha * (W @ phi) (because alpha is multiplied with a matrix in the former while it is multiplied with a vector in the later).

I think `jnp.expand_dims(self.alpha, -1)` now that the multiplication by `alpha` is done after `W @ phi`, so I don't think we need to create `jnp.expand_dims(alpha, (-2, -1))`.